src/Radio/APRS/APRSMsgBase.cpp

```diff
@@ -0,0 +1,104 @@
+#include "APRSMsgBase.h"
+
+APRSMsgBase::APRSMsgBase(APRSConfig header)
+{
+    this->header = header;
+}
+
+bool APRSMsgBase::encode()
+{
+    int c = encodeHeader();
+    encodeData(c);
+    return true;
+}
+
+bool APRSMsgBase::decode()
+{
+    // message needs to be decoded and values reinserted into the header and data structs
+    int c = decodeHeader();
+    decodeData(c);
+    return true;
+}
+
+#pragma region APRSMsgBase Header Encoding/Decoding
+
+int APRSMsgBase::encodeHeader()
+{
+    // format: CALLSIGN>TOCALL,PATH:
+    int cursor = 0;
+
+    for (int i = 0; header.CALLSIGN[i] && i < 8; i++, cursor++)
+        string[cursor] = header.CALLSIGN[i];
+
+    string[cursor++] = '>';
+
+    for (int i = 0; header.TOCALL[i] && i < 8; i++, cursor++)
+        string[cursor] = header.TOCALL[i];
+
+    string[cursor++] = ',';
+
+    for (int i = 0; header.PATH[i] && i < 10; i++, cursor++)
+        string[cursor] = header.PATH[i];
+
+    string[cursor++] = ':'; // end of header
+    return cursor;
+}
+
+int APRSMsgBase::decodeHeader()
+{
+    // format: CALLSIGN>TOCALL,PATH:
+    int cursor = 0;
+
+    for (int i = 0; string[cursor] != '>'; i++, cursor++)
+        header.CALLSIGN[i] = string[cursor];
+    header.CALLSIGN[cursor] = '\0';
+
+    cursor++; // skip '>'
+
+    for (int i = 0; string[cursor] != ','; i++, cursor++)
+        header.TOCALL[i] = string[cursor];
+    header.TOCALL[cursor] = '\0';
+
+    cursor++; // skip ','
+
+    for (int i = 0; string[cursor] != ':'; i++, cursor++)
+        header.PATH[i] = string[cursor];
+    header.PATH[cursor] = '\0';
+
+    cursor++; // skip ':'
+    return cursor;
+}
+
+#pragma endregion
+
+#pragma region Base91 Encoding
+
+void APRSMsgBase::encodeBase91(uint8_t *message, int &cursor, int value, int precision)
+{
+    for (int i = precision - 1; i >= 0; i--)
+    {
+        int divisor = pow(91, i);
+        message[cursor++] = (uint8_t)((int)(value / divisor) + 33);
+        value %= divisor;
+    }
+}
+
+void APRSMsgBase::decodeBase91(const uint8_t *message, int &cursor, double &value, int precision)
+{
+    value = 0;
+    for (int i = 0; i < precision; i++)
+    {
+        value += (message[cursor++] - 33) * pow(91, precision - i - 1);
+    }
+}
+
+void APRSMsgBase::decodeBase91(const uint8_t *message, int &cursor, int &value, int precision)
```

Comment:
I don't think you need an overloaded method here since C++ has implicit type conversions from ints to doubles

Reply:
actually since it's a pointer you may, but I'd double check that

Reply:
yeah if I wrote this, I'm not sure why I did. My best guess is ease-of-use. Don't have to worry about casting or whatever because it just uses the overloaded version. It may also help with higher precision functions that actually support precision into the decimal places.

Reply:
I think I wrote it when I was trying to figure out how to interface Drew's functions with mine. There were a few values (like lat/long or something) where I didn't want to lose the precision of casting to a int before doing the math within the function, and C++ was giving me an error before I overloaded the function.